AMDGPU: Use ConstantPool as source value for DAG lowered kernarg loads

arsenm opened the pull request:
This isn't quite a constant pool, but probably close enough for this
purpose. We just need some known invariant value address. The aliasing
queries against the real kernarg base pointer will falsely report
no aliasing, but for invariant memory it probably doesn't matter.

arsenm committed:
AMDGPU: Use ConstantPool as source value for DAG lowered kernarg loads

This isn't quite a constant pool, but probably close enough for this
purpose. We just need some known invariant value address. The aliasing
queries against the real kernarg base pointer will falsely report
no aliasing, but for invariant memory it probably doesn't matter.

nhaehnle reviewed (approved):
Seems reasonable to me.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -2321,14 +2322,15 @@ Register AMDGPULegalizerInfo::getSegmentAperture(
   // implicit kernargs.
   if (AMDGPU::getAMDHSACodeObjectVersion(*MF.getFunction().getParent()) >=
       AMDGPU::AMDHSA_COV5) {
+    // TODO: can we be smarter about machine pointer info?
```
```suggestion
```

arsenm committed:
Remove comment

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -414,7 +415,10 @@ void AMDGPUCallLowering::lowerParameter(MachineIRBuilder &B, ArgInfo &OrigArg,
+
+  // This isn't really a constant pool but close enough.
+  MachinePointerInfo PtrInfo(MF.getPSVManager().getConstantPool(),
+                             AMDGPUAS::CONSTANT_ADDRESS);
```
```suggestion
  MachinePointerInfo PtrInfo = getKernargSegmentPtrInfo(B.getMF());
```

arsenm committed:
Use helper

arsenm committed:
Fix build

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -414,7 +415,8 @@ void AMDGPUCallLowering::lowerParameter(MachineIRBuilder &B, ArgInfo &OrigArg,
   const DataLayout &DL = F.getDataLayout();
-  MachinePointerInfo PtrInfo(AMDGPUAS::CONSTANT_ADDRESS);
+  const SITargetLowering &TLI = *getTLI<SITargetLowering>();
+  MachinePointerInfo PtrInfo = TLI.getKernargSegmentPtrInfo(B.getMF());
```
```suggestion
  MachinePointerInfo PtrInfo = TLI.getKernargSegmentPtrInfo(MF);
```

arsenm committed:
Use MF variable

arsenm committed:
AMDGPU: Fix treating unknown mem operands as uniform (#168980)

The test changes are mostly GlobalISel specific regressions.
GlobalISel is still relying on isUniformMMO, but it doesn't really
have an excuse for doing so. These should be avoidable with new
regbankselect.

There is an additional regression for addrspacecast for cov4. We
probably ought to be using a separate PseudoSourceValue for the
access of the queue pointer.

arsenm committed:
Revert "AMDGPU: Fix treating unknown mem operands as uniform (#168980)"

This reverts commit d23e1765a9a7cc52673e374be7869f5f0ffc6486.

arsenm committed:
Fix build

arsenm committed:
fix test
